src/address.rs

```diff
@@ -94,19 +95,21 @@ impl fmt::String for Mailbox {
 }
 
 impl FromStr for Mailbox {
-    fn from_str(s: &str) -> Option<Mailbox> {
+    type Err = ParsingError;
+
+    fn from_str(s: &str) -> ParsingResult<Mailbox> {
         AddressParser::new(s).parse_mailbox()
     }
 }
 
 impl FromHeader for Vec<Address> {
-    fn from_header(value: String) -> Option<Vec<Address>> {
-        Some(AddressParser::new(value.as_slice()).parse_address_list())
+    fn from_header(value: String) -> ParsingResult<Vec<Address>> {
+        AddressParser::new(value.as_slice()).parse_address_list()
     }
 }
 
 impl ToFoldedHeader for Vec<Address> {
-    fn to_folded_header(start_pos: usize, value: Vec<Address>) -> Option<String> {
+    fn to_folded_header(start_pos: usize, value: Vec<Address>) -> ParsingResult<String> {
```

Comment:
The intent of `to_header` and `to_folded_header` returning an Option was to allow implementors to indicate a failure to stringify their value. This could be implemented using associated types where we allow the definintion of an error type per implementation. Perhaps something like:

```rust
trait ToFoldedHeader {
    type StringifyError;
 
    fn to_folded_header(pos: usize, value: Self) -> Result<String, Self::StringifyError>;
}
 
struct Foo {
    a: usize,
}
 
impl ToFoldedHeader for Foo {
    type StringifyError = ();
 
    fn to_folded_header(pos: usize, value: Self) -> Result<String, ()> {
        return Ok(format!("{}", value.a))
    }
}
 
fn print_folded<T: ToFoldedHeader>(val: T) {
    println!("{}", ToFoldedHeader::to_folded_header(0, val).unwrap_or("FAIL!".to_string()));
}
 
fn main() {
    let a = Foo { a: 10 };
    print_folded(a);
}
``` 

This allows, like in the Foo case, for something to not have an error type (through the use of Unit) or specify their own errors.